core/kernel/tee_ta_manager.c

```diff
@@ -823,6 +823,190 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
+				if (cnt == MAX_DUMP_SESS_NUM)
+					break;
+
+				dump_ctx[index].sess_id[cnt] =
+					sess->id;
+				cnt++;
+			}
+		}
+
+		dump_ctx[index].sess_num = cnt;
+		index++;
+	}
+}
+
+static TEE_Result dump_ta_stats(struct tee_ta_dump_ctx *dump_ctx,
+				struct tee_ta_dump_stats *dump_stats,
+				uint32_t user_ta_num)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_session *sess = NULL;
+	struct tee_ta_session_head *open_sessions = NULL;
+	struct tee_ta_param param = { };
+
+	nsec_sessions_list_head(&open_sessions);
+
+	for (uint32_t i = 0; i < user_ta_num; i++) {
+		struct tee_ta_dump_stats *stats = &dump_stats[i];
+
+		memcpy(&stats->uuid, &dump_ctx[i].uuid,
+		       sizeof(dump_ctx[i].uuid));
+		stats->panicked = dump_ctx[i].panicked;
+		stats->sess_num = dump_ctx[i].sess_num;
+
+		/*find a session from dump context*/
+		sess = NULL;
+		for (uint32_t j = 0; j < dump_ctx[i].sess_num; j++) {
+			uint32_t sess_id = dump_ctx[i].sess_id[j];
+
+			sess = tee_ta_get_session(sess_id,
+						  true,
+						  open_sessions);
+			if (sess)
+				break;
+		}
+
+		if (!sess)
+			continue;
+		/*if session is existing, get its heap stats*/
+		memset(&param, 0, sizeof(struct tee_ta_param));
+		param.types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+					      TEE_PARAM_TYPE_VALUE_OUTPUT,
+					      TEE_PARAM_TYPE_VALUE_OUTPUT,
+					      TEE_PARAM_TYPE_NONE);
+		res = dump_ta_memstats(sess, TEE_TIMEOUT_INFINITE, &param);
+		if (res == TEE_SUCCESS) {
+			stats->heap.allocated = param.u[0].val.a;
+			stats->heap.max_allocated = param.u[0].val.b;
+			stats->heap.size = param.u[1].val.a;
+			stats->heap.num_alloc_fail = param.u[1].val.b;
+			stats->heap.biggest_alloc_fail = param.u[2].val.a;
+			stats->heap.biggest_alloc_fail_used = param.u[2].val.b;
+		}
+		tee_ta_put_session(sess);
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result tee_ta_instance_stats(void *buff,
+				 uint32_t *buff_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_ta_dump_stats *dump_stats = NULL;
+	struct tee_ta_dump_ctx *dump_ctx = NULL;
+	struct tee_ta_ctx *ctx = NULL;
+	uint32_t actual_size = 0;
+	uint32_t user_ta_num = 0;
+
+	if (!buff_size)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&tee_ta_mutex);
+
+	/*
+	 * Go through all available TA and calc out the actual buffer size.
+	 */
+	TAILQ_FOREACH(ctx, &tee_ctxes, link)
+		if (is_user_ta_ctx(&ctx->ts_ctx))
+			user_ta_num++;
+
+	actual_size = sizeof(struct tee_ta_dump_stats) * user_ta_num;
+	if (!buff || *buff_size < actual_size) {
+		/*
+		 * Buff is null or pass size less than actual size
+		 * means caller try to query the buffer size.
+		 * update *buff_size and return TEE_ERROR_SHORT_BUFFER
+		 */
+		*buff_size = actual_size;
+		res = TEE_ERROR_SHORT_BUFFER;
+		mutex_unlock(&tee_ta_mutex);
+		return res;
+	}
+
+	dump_stats = (struct tee_ta_dump_stats *)buff;
+	actual_size = user_ta_num * sizeof(struct tee_ta_dump_ctx);
+	dump_ctx = (struct tee_ta_dump_ctx *)malloc(actual_size);
```

Comment:
No need to cast the result of `malloc()`.
What if `malloc()` fails?

Reply:
done